### PR TITLE
[WIP] Aten fx pad

### PR DESCRIPTION
**Pull request by msaroufim**
Fixes #ISSUE_NUMBER

Note to self on perf benchmarks, use the local script so something like `benchmarks/dynamo/timm_models.py  —performance --inductor --device cuda --training --amp --only dla102` and then when im ready for the full expensive perf suite I can run`ciflow/inductor-perf-compare`

To debug syntax issues in fx graph I can use `print(graphmodule.code)`

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -7151,6 +7153,40 @@ def foo(m, inp):
             self.assertTrue(
                 foo(m, inp)[0].is_contiguous(memory_format=torch.channels_last)
             )
+        
+        # So far these 2 tests just assert that nothing broke
+        # But ideally I'd like a a test that checks for whether a string is present in the IR
+        # I wonder if there's a better way to do this
+        @requires_cuda()
```
Comment (msaroufim) [inline test/inductor/test_torchinductor.py:7160]: @Chillee any feedback on what's a good test setup - in my notebook what i was doing to validate things was
1. `gm.print_readable()` and making sure new padding op and slicing ops are there
2. `gm.forward()` still works with the same input shape
3. I could also quickly validate that it makes things faster

And I figure to test 
1. I can either test that the compiled model  works with shape padding so more e2e - i couldnt find any existing tests for the existign shape padding implementation
2. And I can test the pass in a standalone way using functorch

The fx/passes directory didn't have enough passes for me to justify putting things there

File: torch/_inductor/padding.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+def get_alignment_size(dtype):
+    # try:
+    #     gpu_name = torch.cuda.get_device_name()
+    # except RuntimeError:
+    #     gpu_name = ""
+    # if "A100" in gpu_name:
+    #     has_a100 = True
+    # else:
+    #     has_a100 = False
+    if dtype == torch.int8:
```
Comment (msaroufim) [inline torch/_inductor/padding.py:28]: This is 128 / dtype heuristic

File: torch/_inductor/padding.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import torch
+
+def needs_padding(args):
+  
+  for arg in args:
+    size = arg.meta["tensor_meta"].shape
+    for i in size:
+      if i // get_alignment_size(i) or i < 512:
```
Comment (msaroufim) [inline torch/_inductor/padding.py:8]: Looking at hte charts you developed 512 seemed to be where padding starts to matter

File: torch/_inductor/decomposition.py [hunk 4]
```diff
@@ -74,225 +74,22 @@ def empty_permuted(size, physical_layout, **kwargs):
...
-def should_pad_bench(mat1, mat2, op, input=None):
-    assert utils.has_triton()
-    from triton.testing import do_bench
-
-    with no_dispatch():
-        if op is torch.ops.aten.mm or op is torch.ops.aten.addmm:
-            m_padded_length = get_padded_length(mat1.shape[0], get_alignment_size(mat1))
-            k_padded_length = get_padded_length(mat1.shape[1], get_alignment_size(mat1))
-            n_padded_length = get_padded_length(mat2.shape[1], get_alignment_size(mat2))
-        elif op is torch.ops.aten.bmm:
```
Comment (msaroufim) [inline torch/_inductor/decomposition.py:157]: For now I only handle `addmm` but will need to also look at `mm` and `bmm`

At least for linear layers what i seemed to be getting was `addmm`

File: torch/_inductor/padding.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    if dtype == torch.int8:
+        return 16
+    elif dtype in [torch.float16, torch.half, torch.bfloat16]:
+        return 8
+    elif dtype == torch.float32:
+        return 4
+    elif dtype == torch.float64:
+        return 2
+    else:
+        return 0
```
Comment (msaroufim) [inline torch/_inductor/padding.py:37]: CI crashing because of this 0

File: torch/_inductor/padding.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    # Leverages a classic interpreter pattern, thanks Horace!
+    new_graph = torch.fx.Graph()
+    env = {}
+    for node in fx_g.graph.nodes:
+        if node.target == torch.ops.aten.addmm.default:
+          
+          # Currently this is a heuristic that decides if we should pad
+          # Decided to only pad for medium size matrices and if alignment is off
+          if needs_padding(node.args):
+              size = int(tuple(env[node.args[0]].meta["tensor_meta"].shape)[0])
```
Comment (msaroufim) [inline torch/_inductor/padding.py:50]: Need to generalize this over all dims

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (Chillee) [reply test/inductor/test_torchinductor.py:7160]: Yeah that's mostly fine.

File: torch/_inductor/decomposition.py (hunk 4 above)
Comment (Chillee) [reply torch/_inductor/decomposition.py:157]: Probably shouldn't delete these until your pass actually works for all these operators.

File: torch/_inductor/padding.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import torch
+
+def needs_padding(args):
+  
+  for arg in args:
+    size = arg.meta["tensor_meta"].shape
```
Comment (Chillee) [inline torch/_inductor/padding.py:6]: Don't use `meta['tensor_meta'].shape` - use `arg.meta['val'].shape`

File: torch/_inductor/padding.py (hunk 3 above)
Comment (Chillee) [inline torch/_inductor/padding.py:8]: Should probably be `i % get_alignment_size`

File: torch/_inductor/padding.py (hunk 5 above)
Comment (Chillee) [reply torch/_inductor/padding.py:37]: Should be `1`.

File: torch/_inductor/padding.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+          # Decided to only pad for medium size matrices and if alignment is off
+          if needs_padding(node.args):
+              size = int(tuple(env[node.args[0]].meta["tensor_meta"].shape)[0])
+              alignment = get_alignment_size(env[node.args[0]].meta["tensor_meta"].dtype)
+              pad_amount = larger_closest_multiple(size, alignment) - size
+
+              # For each matmul, pad the matrix with zeroes, do the matmu and then slice to return a tensor size that the user expects
+              new_a_pad = new_graph.call_function(torch.ops.aten.cat, (env[node.args[0]], torch.ops.aten.zeros([pad_amount, 1])))
+              new_mm_pad = new_graph.call_function(torch.ops.aten.addmm.default, (new_a_pad, env[node.args[1]]))
+              new_mm = new_graph.call_function(torch.ops.aten.slice, (new_mm_pad, 0, pad_amount))
```
Comment (Chillee) [inline torch/_inductor/padding.py:57]: This isn't correct.

**Review by Chillee — CHANGES_REQUESTED** [review]
I don't think this current impl works. Have you tried evaluating this on any torchbench models?

**Comment by msaroufim** [comment]
Closing for now, might revisit in the future
